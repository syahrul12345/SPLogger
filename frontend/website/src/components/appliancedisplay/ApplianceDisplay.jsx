import React from "react";

import { Grid, Card, CardContent, Typography, Button } from "@material-ui/core";

// Redux
import { connect } from "react-redux";
import { UpdateAppliances } from "../../redux-modules/user/actions";

// API calls
import { SubmitDelete } from "../../utils/utils";

// Components
import { ConnectedEditDialog } from "../editDialog";

const ApplianceDisplay = (props) => {
  const { appliances } = props;
  const deleteCallback = (response, deletedAppliance) => {
    const {
      data: { Status },
    } = response;
    // Succesfully delete
    if (Status === true) {
      const newAppliances = appliances.filter(
        (appliance) => appliance.ID !== deletedAppliance.ID
      );
      props.dispatch(UpdateAppliances(newAppliances));
    }
  };
  const deleteFailCallback = (err) => {
    console.log("failed");
    console.log(err);
  };
  const deleteAppliance = (appliance) => {
    SubmitDelete(deleteCallback, deleteFailCallback, appliance);
  };
  return (
    <Grid container spacing={2}>
      {appliances &&
        appliances.map((appliance) => {
          return (
            <Grid item key={appliance.ID} xs={12} md={4}>
              <Card>
                <CardContent>
                  <Typography variant="h5">Brand: {appliance.Brand}</Typography>
                  <Typography variant="h5">Model: {appliance.Model}</Typography>
                  <Typography variant="body1">
                    Status: {appliance.Status}
                  </Typography>
                  <Typography variant="subtitle2">
                    SerialNumber: {appliance.SerialNumber}
                  </Typography>
                  <Typography variant="subtitle2">
                    Date Bought: {appliance.DateBought}
                  </Typography>
                  <Typography variant="subtitle2">
                    Updated At: {appliance.CreatedAt}
                  </Typography>
                </CardContent>
                <Grid
                  container
                  style={{ paddingLeft: "1%", paddingBottom: "2%" }}
                  spacing={3}
                  justify="space-between"
                >
                  <Grid item xs={3}>
                    <ConnectedEditDialog selectedAppliance={appliance} />
                  </Grid>
                  <Grid item xs={3}>
                    <Button onClick={() => deleteAppliance(appliance)}>
                      {" "}
                      Delete
                    </Button>
                  </Grid>
                </Grid>
              </Card>
            </Grid>
          );
        })}
    </Grid>
  );
};

const mapStateToProps = (state) => {
  return {
    appliances: state.user.appliances,
  };
};
const ConnectApplianceDisplay = connect(mapStateToProps)(ApplianceDisplay);

export { ApplianceDisplay, ConnectApplianceDisplay };
