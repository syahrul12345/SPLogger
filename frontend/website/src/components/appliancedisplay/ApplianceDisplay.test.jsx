import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";

import { ApplianceDisplay } from "./ApplianceDisplay";
import { Card, CardContent, Typography } from "@material-ui/core";

// Mock the redux state
const mockAppliances = [
  {
    ID: 22,
    CreatedAt: "2020-06-26T07:51:48.306436Z",
    UpdatedAt: "2020-06-26T07:51:48.306436Z",
    SerialNumber: "CH-1",
    Brand: "Starbucks",
    Status: "New",
    Model: "Cup",
    DateBought: "Today",
  },
  {
    ID: 25,
    CreatedAt: "2020-06-26T07:52:29.048139Z",
    UpdatedAt: "2020-06-26T07:52:29.048139Z",
    SerialNumber: "CH-2",
    Brand: "CBTL",
    Status: "New",
    Model: "Tumblr",
    DateBought: "Today",
  },
];

// Define the tests
describe("Render tests for ApplianceDispaly", () => {
  test("Comparing snapshot", () => {
    const component = renderer.create(<ApplianceDisplay />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Test with 2 appliance objects as props", () => {
  const component = shallow(<ApplianceDisplay appliances={mockAppliances} />);
  test("Test there are 2 cards present", () => {
    expect(component.find(Card).length).toEqual(2);
  });
  mockAppliances.forEach((mockAppliance, idx) => {
    test(`Find card with Brand field as ${mockAppliance["Brand"]}`, () => {
      expect(
        component.find(Card).at(idx).find(CardContent).childAt(0).text()
      ).toEqual("Brand: " + mockAppliance["Brand"]);
    });
    test(`Find card with Model field as ${mockAppliance["Model"]}`, () => {
      expect(
        component.find(Card).at(idx).find(CardContent).childAt(1).text()
      ).toEqual("Model: " + mockAppliance["Model"]);
    });
    test(`Find card with status as ${mockAppliance["Status"]}`, () => {
      expect(
        component.find(Card).at(idx).find(CardContent).childAt(2).text()
      ).toEqual("Status: " + mockAppliance["Status"]);
    });
    test(`Find card with SerialNumber as ${mockAppliance["SerialNumber"]}`, () => {
      expect(
        component.find(Card).at(idx).find(CardContent).childAt(3).text()
      ).toEqual("SerialNumber: " + mockAppliance["SerialNumber"]);
    });
  });
});
