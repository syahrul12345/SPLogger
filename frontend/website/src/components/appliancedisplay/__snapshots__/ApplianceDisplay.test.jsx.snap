// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render tests for ApplianceDispaly Comparing snapshot 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
/>
`;
